test/websocket.js

```diff
@@ -0,0 +1,121 @@
+const assert = require("assert");
+const nodemailer = require("nodemailer");
+const request = require("supertest");
+const WebSocket = require("ws");
+
+const mailsacAPIKey = ""; // Generated by mailsac. See https://mailsac.com/api-keys
+const mailsacToAddress = ""; // Mailsac email address where the email will be sent
+const smtpUserName = ""; // Username for smtp server authentication
+const smtpPassword = ""; // Password for smtp server authentication
+const smtpHost = ""; // hostname of the smtp server
+const smtpPort = 587; // port the smtp is listening on
+
+let ws;
+
+describe("send email to mailsac", function () {
+  this.timeout(50000); // test can take a long time to run. This increases the default timeout for mocha
+
+  /* delete all messages in the inbox after the test runs to prevent leaky tests.
+       This requires the inbox to private, which is a paid feature of Mailsac.
+       The afterEach section could be omitted if using a public address
+    */
+  afterEach(() =>
+    request("https://mailsac.com")
+      .delete(`/api/addresses/${mailsacToAddress}/messages`)
+      .set("Mailsac-Key", mailsacAPIKey)
+      .expect(204)
+  );
+  // Open websocket waiting for email. This websocket will be reused for tests in this file.
+  before(() => {
+    ws = new WebSocket(
+      `wss://sock.mailsac.com/incoming-messages?key=${mailsacAPIKey}&addresses=${mailsacToAddress}`
+    );
+    ws.on("open", () => {
+      console.log("web socket opened");
+    });
+    ws.on("error", (err) => {
+      console.log("connection error", err);
```

Comment:
It seems like these event handlers need to be wrapped in a promise returned to `before`. Otherwise the `it()` blocks are going to run before we know whether the web socket connection worked.

Reply:
Perhaps `ws` is buffering requests internally until the connection happens, so it's working. But it reads out-of-order to me.

Reply:
I'm not sure what you mean by "out of order". Could you explain in more detail what the problem is and how you would suggest solving it?